### PR TITLE
Fix a slew of fuzz bugs in wit-{parser,component}

### PR DESCRIPTION
**Pull request by alexcrichton**
This PR fixes the `roundtrip-wit` fuzz target to actually be able to generate `use` statements. An accidental bug meant that the fuzzer previously didn't actually create `use` statements, despite that being one of the main goals of the fuzzer. Alas! 

In any case this turned up a number of bugs and additionally found https://github.com/bytecodealliance/wasm-tools/pull/885 in no time at all.

### CODE REVIEW

**Comment by alexcrichton** [comment]
Note that this is a draft while the fuzzer is still running as I've got more bugs to fix, and I'll keep pushing them here as I discover them.

**Comment by alexcrichton** [comment]
Ok there ended up being quite a few more issues than I thought there were going to be. @peterhuene I think this is ready to go as I've got 70 cores of fuzzing being clean for ~30 minutes now. I've tried to split everything up where each commit fixes an individual problem and a single fuzz test case. As always though I'm happy to dive into more details with you or expand comments and such.

File: fuzz/fuzz_targets/roundtrip-wit.rs [hunk 1]
```diff
@@ -462,7 +502,6 @@ mod generate {
                 ret.push_str("\n\n");
             }
 
-            self.types_in_interface.clear();
```
Comment (peterhuene) [inline fuzz/fuzz_targets/roundtrip-wit.rs:465]: Whoops, sorry for not catching that either!

**Review by peterhuene — APPROVED** [review]
Looks 👍. Just minor nits.

File: fuzz/fuzz_targets/roundtrip-wit.rs (hunk 1 above)
Comment (alexcrichton) [reply fuzz/fuzz_targets/roundtrip-wit.rs:465]: Heh I'm the one who ran the fuzzer for hours without noticing this 🤡

**Pull request merged**